packages/terminal/src/node/base-terminal-server.ts

```diff
@@ -68,6 +68,27 @@ export abstract class BaseTerminalServer implements IBaseTerminalServer {
         }
     }
 
+    async onAttached(id: number): Promise<void> {
+        const terminal = this.processManager.get(id);
+        if (terminal instanceof TaskTerminalProcess) {
+            terminal.attached = true;
+            if (terminal.exited) {
+                // Didn't execute `unregisterProcess` on terminal `exit` emitted to enable attaching task output,
+                // Fixes issue [#2961](https://github.com/eclipse-theia/theia/issues/2961).
+                terminal.unregisterProcess();
+            }
+        }
+    }
+
+    async onFailedAttach(id: number): Promise<void> {
+        const terminal = this.processManager.get(id);
+        if (terminal instanceof TaskTerminalProcess) {
+            if (terminal.exited) {
+                terminal.unregisterProcess();
```

Comment:
It looks like this logic could cause leave some processes hanging. If an attachment attempt fails but `terminal.exited` is false, this won't unregister the process - will it ever be unregistered?

Reply:
I should set `terminal.attached = true` and so it will be unregister on exit... (in tastTerminalProcess)
but the naming not suit, WDYS?

Reply:
Hm... so I think if we want to get around the naming, we could call the flag 'attachmentAttempted' or something. Then the guarantee is that the process won't disappear before an attachment is attempted, but there's no guarantee that you can retry. And the frontend code _doesn't_ retry, so that's fine, as a practical matter.

Really, I think that a purely back-end solution is always going to have to accept some sort of compromise like that, so I think it's fine for now. Ideally, I think it would be nice to have the following flow:
```
Frontend creates a terminal
----> Sends information about that terminal to the backend
        -----> Backend creates process and hooks process up to terminal specified by front end.
```
That way we avoid this problem entirely: all of the frontend work is done by the time the process is created, so there's no chance of the process dying before the frontend can attach. But none of the backend code is set up to (conditionally) accept that kind of input from the frontend, so it would take quite a bit of work to refactor everything appropriately, so it's likely beyond the scope of this PR - unless you want to take it on.

So in the meantime, I'm satisfied if you just try to ensure that no process will be left hanging.

Reply:
> I would have been happy to do so, Looks much better design... but do not think I can do that, So I updated the code to have the missing flag & renamed it, Is that all or I missed something?

I think you've covered all cases, now. It looks like, though that the code for `onAttach` and `onFailedAttach` is the same, so maybe they could both just be `onAttachAttempted`?

Reply:
Absolutely right! updated...